sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/action/DetectNewPartitionsAction.java

```diff
@@ -146,15 +146,22 @@ private ProcessContinuation schedulePartitions(
       OutputReceiver<PartitionMetadata> receiver,
       Timestamp minWatermark,
       TreeMap<Timestamp, List<PartitionMetadata>> batches) {
+    List<PartitionMetadata> batchPartitionsDifferentCreatedAt = new ArrayList<>();
+    int numTimestampsHandledSofar = 0;
     for (Map.Entry<Timestamp, List<PartitionMetadata>> batch : batches.entrySet()) {
+      numTimestampsHandledSofar++;
       final Timestamp batchCreatedAt = batch.getKey();
-      final List<PartitionMetadata> batchPartitions = batch.getValue();
-
-      final Timestamp scheduledAt = updateBatchToScheduled(batchPartitions);
-      if (!tracker.tryClaim(batchCreatedAt)) {
-        return ProcessContinuation.stop();
+      final List<PartitionMetadata> batchPartitionsSameCreatedAt = batch.getValue();
+      batchPartitionsDifferentCreatedAt.addAll(batchPartitionsSameCreatedAt);
+      if (batchPartitionsDifferentCreatedAt.size() >= 200
+          || numTimestampsHandledSofar == batches.size()) {
+        final Timestamp scheduledAt = updateBatchToScheduled(batchPartitionsDifferentCreatedAt);
+        if (!tracker.tryClaim(batchCreatedAt)) {
+          return ProcessContinuation.stop();
+        }
+        outputBatch(receiver, minWatermark, batchPartitionsDifferentCreatedAt, scheduledAt);
+        batchPartitionsDifferentCreatedAt = new ArrayList<>();
```

Comment:
I think you could also call the clear() function?

Reply:
Do you mean call batchPartitionsDifferentCreatedAt.clear() instead of initializing a new ArrayList at line 163? I thought clear() has time complexity O(n) while initialize new ArrayList has O(1) but clear() may be better from other considerations. Let me know if you still prefer clear().